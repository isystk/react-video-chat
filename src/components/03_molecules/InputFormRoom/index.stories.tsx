import { Meta, Story } from '@storybook/react'
import React from 'react'
import InputFormRoom from './index'
import MainService from '@/services/main'

export default {
  title: '03_molecules/InputFormRoom',
  component: InputFormRoom,
} as Meta

const main = {
  self: {
    name: 'isystk',
  },
  room: {
    name: '',
  },
} as MainService

const Template: Story = () => <InputFormRoom main={main} />

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
