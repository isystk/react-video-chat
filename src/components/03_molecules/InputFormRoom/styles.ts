import { css, keyframes } from '@emotion/css'

const useStyles = (theme) => css`{
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
  form: {
    margin: '40px auto',
    maxWidth: '500px',
  },
}`

export { useStyles }
