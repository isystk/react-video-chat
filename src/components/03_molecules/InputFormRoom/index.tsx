import React, { FC, useContext, useEffect, useState } from 'react'
import Button from '@material-ui/core/Button'
import CssBaseline from '@material-ui/core/CssBaseline'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import Container from '@material-ui/core/Container'
import { ContainerProps, WithChildren } from 'types'
import { useStyles } from './styles'
import { connect } from '@/components/hoc'
import MainService from '@/services/main'
import { Context } from '@/components/05_layouts/HtmlSkeleton'

/** InputFormRoomProps Props */
export type InputFormRoomProps = WithChildren
/** Presenter Props */
export type PresenterProps = InputFormRoomProps & {
  main
  classes
  label
  name
  setName
  isComposed
  setIsComposed
  initializeRemotePeer
  disabled
}

/** Presenter Component */
const InputFormRoomPresenter: FC<PresenterProps> = ({
  main,
  classes,
  label,
  name,
  setName,
  isComposed,
  setIsComposed,
  initializeRemotePeer,
  disabled,
  ...props
}) => (
  <>
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <div className={classes.paper}>
        <Typography component="h1" variant="h5">
          {label}を入力してください
        </Typography>
        <form className={classes.form} noValidate>
          <TextField
            autoFocus
            fullWidth
            label={label}
            margin="normal"
            name="name"
            onChange={(e) => setName(e.target.value)}
            onCompositionEnd={() => setIsComposed(false)}
            onCompositionStart={() => setIsComposed(true)}
            onKeyDown={async (e) => {
              if (isComposed) return
              if (e.target.value === '') return
              if (e.key === 'Enter') await initializeRemotePeer(e)
            }}
            required
            value={name}
            variant="outlined"
          />
          <Button
            className={classes.submit}
            color="primary"
            disabled={disabled}
            fullWidth
            onClick={async (e) => await initializeRemotePeer(e)}
            type="submit"
            variant="contained"
          >
            決定
          </Button>
        </form>
      </div>
    </Container>
  </>
)

/** Container Component */
const InputFormRoomContainer: React.FC<
  ContainerProps<InputFormRoomProps, PresenterProps>
> = ({ presenter, children, ...props }) => {
  const main = useContext<MainService | null>(Context)
  if (!main) return <></>
  const classes = useStyles()
  const label = '部屋の名前'
  const [disabled, setDisabled] = useState(true)
  const [name, setName] = useState('')
  const [isComposed, setIsComposed] = useState(false)

  useEffect(() => {
    const disabled = name === ''
    setDisabled(disabled)
  }, [name])

  const initializeRemotePeer = async (
    e:
      | React.KeyboardEvent<HTMLDivElement>
      | React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.persist()

    if (name !== '') {
      await main.setRoomName(name)
    }

    e.preventDefault()
  }

  if (main.self.name === '') return <></>
  if (main.room.name !== '') return <></>

  return presenter({
    children,
    main,
    classes,
    label,
    name,
    setName,
    isComposed,
    setIsComposed,
    initializeRemotePeer,
    disabled,
    ...props,
  })
}

export default connect<InputFormRoomProps, PresenterProps>(
  'InputFormRoom',
  InputFormRoomPresenter,
  InputFormRoomContainer
)
