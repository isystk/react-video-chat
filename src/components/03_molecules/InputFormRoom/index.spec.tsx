import React, { useState } from 'react'
import renderer from 'react-test-renderer'
import InputFormRoom from './index'
import '@testing-library/jest-dom/extend-expect'
import { renderHook } from '@testing-library/react-hooks'
import Main from '@/services/main'

describe('InputFormRoom', () => {
  it('Match Snapshot', () => {
    const stateMain = renderHook(() => useState<Main | null>(null))
    const [, setAppRoot] = stateMain.result.current
    const main = new Main(setAppRoot)
    if (main === null) return
    const component = renderer.create(<InputFormRoom main={main} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
