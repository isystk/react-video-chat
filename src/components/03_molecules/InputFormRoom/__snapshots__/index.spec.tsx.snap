// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputFormRoom Match Snapshot 1`] = `null`;
